### PR TITLE
cache for states

### PR DESCRIPTION
**Pull request by cendhu**
Signed-off-by: senthil <cendhu@gmail.com>

### CODE REVIEW

File: internal/worldstate/leveldb/cache.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	return nil
+}
+
+// putStateIfExist stores a given value in the cache only if the key already
+// exists in the cache
+func (c *cache) putStateIfExist(namespace, key string, cacheValue []byte) {
+	cacheKey := constructCacheKey(namespace, key)
+
+	if c.dataCache.Has(cacheKey) {
+		fmt.Println(namespace, key)
```
Comment (liran-funaro) [inline internal/worldstate/leveldb/cache.go:65]: Please remove debugging code.

**Review by liran-funaro — CHANGES_REQUESTED** [review]

File: internal/worldstate/leveldb/cache.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+// putState stores a given value in the cache
+func (c *cache) putState(namespace, key string, cacheValue []byte) error {
+	cacheKey := constructCacheKey(namespace, key)
+
+	if c.dataCache.Has(cacheKey) {
+		c.dataCache.Del(cacheKey)
+	}
+
+	c.dataCache.Set(cacheKey, cacheValue)
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/worldstate/leveldb/cache.go:53]: When the length of the value cannot be encoded within 2 bytes or does not fit into a chunk, the value is not updated. Hence, we always delete before updating the value. We have a similar logic in fabric too while using this cache library - https://github.com/hyperledger/fabric/blob/main/core/ledger/kvledger/txmgmt/statedb/statecouchdb/cache.go

File: internal/worldstate/leveldb/cache.go (hunk 1 above)
Comment (cendhu) [reply internal/worldstate/leveldb/cache.go]: Done

File: internal/worldstate/leveldb/cache.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+// putStateIfExist stores a given value in the cache only if the key already
+// exists in the cache
+func (c *cache) putStateIfExist(namespace, key string, cacheValue []byte) {
+	cacheKey := constructCacheKey(namespace, key)
+
+	if c.dataCache.Has(cacheKey) {
+		fmt.Println(namespace, key)
+		c.dataCache.Del(cacheKey)
+		c.dataCache.Set(cacheKey, cacheValue)
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/worldstate/leveldb/cache.go:64]: Due to value size limit.

File: internal/worldstate/leveldb/cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liran-funaro) [reply internal/worldstate/leveldb/cache.go:53]: Can you please add a comment explaining this?

[the remark this answers is not in the document]
Comment (liran-funaro) [reply internal/worldstate/leveldb/cache.go:53]: Please address this.

**Pull request merged**